src/dune_pkg/opam_solver.ml

```diff
@@ -60,83 +50,215 @@ end = struct
-  include CONTEXT
+  let is_variable_sys variable =
+    Solver_env.Sys_var.of_string_opt (OpamVariable.to_string variable)
+    |> Option.is_some
 
-  val create :
-       base_context:Base_context.t
-    -> local_packages:OpamFile.OPAM.t OpamPackage.Name.Map.t
-    -> solver_env:Solver_env.t
-    -> t
-end = struct
+  let resolve_sys_env_variables sys_var_bindings =
+    map_filters_bottom_up ~f:(function
+      | FIdent ([], variable, None) as filter -> (
+        (* System environment variables which are set in the solver environment
+           are substituted for their values here. *)
+        let sys_var_value_opt =
+          Solver_env.Sys_var.of_string_opt (OpamVariable.to_string variable)
+          |> Option.bind ~f:(Solver_env.Sys_var.Bindings.get sys_var_bindings)
+        in
+        match sys_var_value_opt with
+        | None -> filter
+        | Some sys_var_value -> FString sys_var_value)
+      | ( FOp (FIdent (_, variable, _), _, _)
+        | FOp (_, _, FIdent (_, variable, _)) ) as filter ->
+        (* Comparisons with unset system environment variables resolve to
+           true. This is so that we add dependencies guarded by filters on
+           unset system variables. For example if a package has a
+           linux-only and a macos-only dependency and the user hasn't
+           specified that they only want to solve for a specific os, then
+           we should add both the linux-only and macos-only dependencies
+           to the solution.
+
+           Note that this branch is only followed for unset variables as
+           [map_filters_bottom_up] traverses the formula bottom up, so
+           variables with values in [solver_env] will have been substituted for
+           those values already by the time control gets here.*)
+        if is_variable_sys variable then FBool true else filter
+      | other -> other)
+end
+
+module Context_for_dune = struct
   let local_package_default_version =
     OpamPackage.Version.of_string Lock_dir.Pkg_info.default_version
 
   type t =
-    { base_context : Base_context.t
+    { repo : Opam_repo.t
+    ; version_preference : Version_preference.t
     ; local_packages : OpamFile.OPAM.t OpamPackage.Name.Map.t
     ; solver_env : Solver_env.t
     }
 
-  let create ~base_context ~local_packages ~solver_env =
-    { base_context; local_packages; solver_env }
+  let create ~solver_env ~repo ~local_packages ~version_preference =
+    { repo; version_preference; local_packages; solver_env }
 
-  type rejection = Base_context.rejection
+  type rejection = |
 
-  let pp_rejection = Base_context.pp_rejection
+  let pp_rejection = Fmt.nop
+
+  let opam_version_compare t a b =
+    let comparison_int =
+      OpamPackage.Version.compare (OpamFile.OPAM.version a)
+        (OpamFile.OPAM.version b)
+    in
+    match t.version_preference with
+    | Oldest -> Ordering.of_int comparison_int
+    | Newest -> Ordering.of_int (-comparison_int)
```

Comment:
I think this code sort of defeats the spirit of having `Ordering` to not deal with integers. Instead, you could immediately convert your `comparison_int` to `Ordering.of_int` and then use a helper function to invert the ordering by pattern matching. Perhaps we could add this helper to `Ordering` itself?

Reply:
Good suggestion. I have added a new function `Ordering.reverse` which reverses the direction of a comparison function and refactored this code to use it.